Custom option for comparing date_range

gmatheus opened the pull request:
Custom date range option/fields for aggregation form. This allows users to compare the global time range filter data with any other date range.

Commit:
Custom option added to comparing offsets array + form

Commit:
Custom comparing handler added to BucketsComparing write function

Commit:
Empty date handler added

Commit:
handleCustomDate function created

Commit:
missing customComparing handler added

Commit:
required validation added to custom comparing fields

Commit:
Form style changed + documentation link added

joao-sallaberry commented on public/agg_types/comparing.js:
```diff
@@ -32,6 +37,30 @@ function getDate(date, offset) {
+
+  // Checks if `customComparing.to` is using day format.
+  //  If so, rounds it to the end of the day
+  const isEndDateUsingTime = customComparing.to && customComparing.to.includes(':');
```
What if the input date is an hour, like `2018/04/21 17`? Will it work?

gmatheus replied:
Unless you explicitly specify this format, the `YYYY/MM/DD HH` pattern is invalid for `moment`, so you'll get an error message in form level, before you can run the query (take a look at the [validate_date_math.js](https://github.com/elastic/kibana/blob/master/src/ui/public/directives/validate_date_math.js) directive used in this field). I know this is not the best solution to check if the string contains time, but I couldn't find another easy way to accomplish that (I wish `moment` had this kind of validation)

joao-sallaberry replied:
That's true but unfortunately `YYYY-MM-DD HH` is valid and has the same issue. Since this is not a priority, I'll open a bug report.

Commit:
Different time ranges warning added in custom comparing form

Commit:
isNullOrEmpty validation added to hasDifferentRanges function

Commit:
customComparingTexts undefined validation added to hasDifferentRanges function

Commit:
Different time ranges warning message changed

Commit:
Plugin version updated to 0.0.2

joao-sallaberry commented on public/agg_types/comparing_controller.js:
```diff
@@ -1,11 +1,37 @@
+    if (!$scope.isCustomComparing() || !customComparingTexts) return false;
+
+    // Checks if input texts are not empty
+    const isNullOrEmpty = text => text == null || text === '' || text && text.trim() === '';
```
`text === ''` is useless. It's already covered by the last clause

gmatheus replied:
In this case, this validation is needed since last clause will return false for empty strings:
`!!'' && ''.trim() === ''` -> `false`

But I think it could be rewritten like this (without `text &&` in last clause, since we're already checking `null` or `empty` in the first one):
`const isNullOrEmpty = text => text == null || text.trim() === '';`

joao-sallaberry replied:
You're right! Great!

joao-sallaberry commented on public/agg_types/lib/custom_date_handler.js:
```diff
@@ -0,0 +1,26 @@
+  //  If so, rounds it to the end of the day
+  const isEndDateUsingTime = customComparing.to && customComparing.to.includes(':');
+  const momentEndDate = moment(customComparing.to);
+  const isEndDateInDayFormat = !isEndDateUsingTime && momentEndDate.isSame(momentEndDate.startOf('day'));
```
great!

Though it does't work with `2018-05-45 00`. It will be rounded to the end of the day when it shouldn't.

Commit:
Cloning end date to get start of day

Cloning date because `.startOf()` actually modifies the date, which is undesired

gmatheus commented on public/agg_types/lib/custom_date_handler.js:
```diff
@@ -16,11 +16,11 @@ export function handleCustomDate(customComparing) {
   const isEndDateUsingTime = customComparing.to && customComparing.to.includes(':');
   const momentEndDate = moment(customComparing.to);
-  const isEndDateInDayFormat = !isEndDateUsingTime && momentEndDate.isSame(momentEndDate.startOf('day'));
+  const isEndDateInDayFormat = !isEndDateUsingTime && momentEndDate.isSame(momentEndDate.clone().startOf('day'));
```
Good one! Thanks for noticing 🙂

Commit:
isNullOrEmpty validation changed in hasDifferentRanges function
